server/channels/app/config.go

```diff
@@ -223,6 +228,18 @@ func (a *App) GetEnvironmentConfig(filter func(reflect.StructField) bool) map[st
 
 // SaveConfig replaces the active configuration, optionally notifying cluster peers.
 func (a *App) SaveConfig(newCfg *model.Config, sendConfigChangeClusterMessage bool) (*model.Config, *model.Config, *model.AppError) {
+	// if Elasticsearch autocomplete was enabled, we need to verify if channels index is in right state.
+	// "Test Connection" button in system console checks this and set a flag in ES service,
+	// so we just need to check that flag's value.
+	oldESChannelAutocompleteWasDisabled := !*a.Config().ElasticsearchSettings.EnableAutocomplete
+	newESChannelAutocompleteIsEnabled := *newCfg.ElasticsearchSettings.EnableAutocomplete
+
+	if oldESChannelAutocompleteWasDisabled && newESChannelAutocompleteIsEnabled {
+		if !a.SearchEngine().ElasticsearchEngine.IsAutocompletionEnabled() {
+			return nil, nil, model.NewAppError("SaveConfig", "app.save_config.elasticsearch.autocomplete_cannot_be_enabled_error", nil, "", http.StatusBadRequest)
+		}
+	}
```

Comment:
I already have a config listener. This check here is to display error message in system console UI on saving, like this-
![Screenshot 2024-02-01 at 3 35 19 PM](https://github.com/mattermost/mattermost/assets/18575143/b28faea0-2950-44de-bd9f-711c553e76b3)

I agree we're not doing this kind of blocking check anywhere else. My rational for doing so was that if I didn't do so, the system bot may go unnoticed and displaying error this was is the quickest way to display error.

If you still think its wrong, I'll check this out.

Reply:
Can we move this inside the "Test connection" logic? Because you cannot save the config until test connection returns true.

Reply:
You  need test connection only when enabling ES or updating the ES URL or credentials. Is ES is already enabled but autocomplete was disabled, then test connection is not required when enabling autocomplete.

Reply:
Oh yes that's right. I think it would be fine to rely on the system bot DM for the notification. I would really prefer to avoid adding such specific blocking checks on config saves. And especially because all of this is just temporary code.

Reply:
Done. Moved to API layer with all other existing checks. Thanks for looking into this!